onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -52,7 +53,191 @@ static void BuildFusedKernelDef(KernelDefBuilder& builder, const IndexedSubGraph
+/// a node with kMSNHWC domain is enough. This is because after compile all the nodes which the EP claims are fused
+/// into 1 and removed from the graph.
+/// </summary>
+/// <param name="graph">Graph to validate</param>
+/// <returns></returns>
+static Status ValidateGraphPartitioning(const Graph& graph) {
+  for (const auto& node : graph.Nodes()) {
+    if (node.Domain() == kMSNHWCDomain) {
+      return Status(common::ONNXRUNTIME, common::FAIL,
+                    "Graph contains an invalid node: " + node.Name() + " Op Type: " + node.OpType() +
+                        " with domain: " + kMSNHWCDomain + ". These are temporary nodes added during layout transformations " +
+                        " and are not expected to remain in the graph post partitioning. This is a bug in layout transformer.");
+    }
+  }
+  return Status::OK();
+}
+
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_ENABLE_RUNTIME_OPTIMIZATION_IN_MINIMAL_BUILD)
+
+/// <summary>
+/// Check if a node can be placed on a specific provider. If yes, then set the nodes execution provider.
+/// Do nothing if the node is already assigned.
+/// </summary>
+/// <param name="graph">Graph in question.</param>
+/// <param name="capability">Indexed subgraph which needs to be assigned</param>
+/// <param name="provider_type">The EP to assign the Indexed subgraph to</param>
+static void AssignNodes(Graph& graph, const IndexedSubGraph& capability,
+                        const std::string& provider_type) {
+  // Before assigning the ep to any node, first walk through all the nodes and ensure
+  // none of the nodes have already been assigned. If a node is assigned, simply return.
+  for (auto node_index : capability.nodes) {
+    const auto* node = graph.GetNode(node_index);
+    if ((nullptr == node) || (!node->GetExecutionProviderType().empty() && node->GetExecutionProviderType() != provider_type)) {
+      return;
+    }
+  }
+
+  for (auto node_index : capability.nodes) {
+    auto* node = graph.GetNode(node_index);
+    if (nullptr != node) {
+      node->SetExecutionProviderType(provider_type);
+    }
+  }
+}
+
+/// <summary>
+/// Transforms data layout from NCHW to NHWC. Applies transforms to layout sensitive nodes
+/// assigned to current_ep and any other non-layout sensitive nodes in order to optimize
+/// the transposes as much as possible.
+/// </summary>
+/// <param name="graph"></param>
+/// <param name="modified"></param>
+/// <param name="current_ep"></param>
+/// <param name="logger"></param>
+/// <returns></returns>
+static Status TransformLayout(Graph& graph, bool& modified,
+                              IExecutionProvider& current_ep) {
+  // sub graph recurse will be added later
+  auto api_graph = MakeApiGraph(graph, current_ep.GetAllocator(0, OrtMemTypeDefault), nullptr);
+  std::unordered_set<std::string_view> layout_sensitive_ops = onnx_layout_transformation::GetLayoutSensitiveOps();
```

Comment:
This is still constructing the set each time as GetLayoutSensitiveOps would need to change to return a reference (with its impl being in a .cc file) to avoid that. 

I guess there's only one call to GetLayoutSensitiveOps so it could be done either way, however using a reference here is slightly misleading in terms of ownership and lifetime as the scope of the returned value is this function